### PR TITLE
Initial PegJS version for parsing chords over words

### LINKED ISSUE #219: 💡 RFC: Parse Metadata in a text chordsheet
Parent Task: https://github.com/PraiseCharts/studio/issues/32
### Background & Motivation

Want to explore the idea of having metadata in a text chord sheet. 

I like the idea of markdown style frontmatter.
```
title: Honey In the Rock
key: F
---
Verse 1
D       Dsus           D       Dsus
 Praying for a miracle, thirsty
```
=
```
{title: Honey In the Rock}
{key: F}
{c:Verse 1}
[D] Praying[Dsus] for a miracle,[D] thirsty[Dsus]
```

### Possible solutions

- add this functionality to the `ChordSheetParser()`
- create a new parser that can do this. `ChordsOverWordsParser()`

### PR DESCRIPTION
**Pull request by martijnversluis**
# TODO/pending:

- [ ] support numerics and numerals (only chord symbols are supported at this stage)
- [x] test with different chord sheets
- [x] the parser assumes a line of text not associated with chords is a comment, but I'm not sure that's always correct
- [x] extract the chord grammar. Add a build step to insert the chord grammar into the chord sheet grammar
- [ ] test with all kinds of chords

# Nice to have, not included in this PR:

- [ ] reuse the chord grammar for `Chord.parse`
- [ ] use the chord grammar in `ChordProParser` to have it parse chords
- [ ] we're parsing chords, but when creating a song we're cramming it into a string because that's what it expects. Instead, we should be able to incorporate the parsed chords into the song. We should either differentiate between a chord object and a chord string, or ensure a chord is always a `Chord` object (but that adds parsing overhead when you might not even need the parsed chord)
- [ ] we're parsing chords, but we're using string keys. Instead, keys (and even notes) should be objects.

Resolves #219

### CODE REVIEW

**Comment by isaiahdahl** [comment]
@martijnversluis Great start at first glance! I can create a bunch more integration tests with various examples.

> - [ ] support numerics and numerals (only chord symbols are supported at this stage)

In the code mirror lezer grammar we started something like this:
```
Chord { symbolChord | numericChord | numeralChord | rhythmSymbol }

symbolChord { ($[A-G] accidental? ($[a-zA-Z0-9\(\#)]*)?) ( '/' $[A-G] accidental?)? }
numericChord { (accidental? $[1-7] ($[a-zA-Z0-9\(\#)]*)?) ( '/' accidental? $[1-7])?  }
numeralChord { (accidental? ('I' | 'II' | 'III' | 'IV' | 'V' | 'VI' | 'VII') ($[a-zA-Z0-9\(\)#]*)?) ( '/' accidental? ('I' | 'II' | 'III' | 'IV' | 'V' | 'VI' | 'VII'))?  }
rhythmSymbol { $[x|/] }
accidental { $[b#] }
```
I'd assume the same sorta concept could work here. Obviously, it's a different syntax. 

> - [ ] the parser assumes a line of text not associated with chords is a comment, but I'm not sure that's always correct

This could definitely have a conflict with lyric only lines in a verse. 
I think a better rule might be something like: a line of text not associated with chords that follows a empty line.

> - [ ] we're parsing chords, but when creating a song we're cramming it into a string because that's what it expects. Instead, we should be able to incorporate the parsed chords into the song. We should either differentiate between a chord object and a chord string, or ensure a chord is always a `Chord` object (but that adds parsing overhead when you might not even need the parsed chord)
> - [ ] we're parsing chords, but we're using string keys. Instead, keys (and even notes) should be objects.
> - [ ] extract the chord grammar and reuse it for `Chord.parse`. Add a build step to insert the chord grammar into the chord sheet grammar
> - [ ] test with all kinds of chords
> - [ ] use the chord grammar in `ChordProParser` to have it parse chords

It sounds like these are things that apply to all parsers, Is that correct?  Maybe we should just focus on getting the features needed for the chords over words parser working and then follow it up with a parser refactor and clean up RFC/task? Thoughts?

**Comment by martijnversluis** [comment]
@isaiahdahl 

> In the code mirror lezer grammar we started something like this:

Really useful! The syntax is different, but I can easily translate the format.

> This could definitely have a conflict with lyric only lines in a verse.
> I think a better rule might be something like: a line of text not associated with chords that follows a empty line, unless it's the first line.

Yeah makes sense! It would be great though to have some real-world examples so I could prove (with a test) that it's not handled correctly and fix it accordingly.

> It sounds like these are things that apply to all parsers, Is that correct? Maybe we should just focus on getting the features needed for the chords over words parser working and then follow it up with a parser refactor and clean up RFC/task? Thoughts?

You're probably right 😄  I guess I got a bit too excited that the grammar works and what possibilities that adds. I'll tag the TODOs I think should be in this PR, and move the rest to a separate issue later.

Testing with different sheets and chords should still be done though.

**Comment by isaiahdahl** [comment]
@martijnversluis 
> Yeah makes sense! It would be great though to have some real-world examples so I could prove (with a test) that it's not handled correctly and fix it accordingly.

I'm currently writing a bunch of tests covering a bunch of edge cases I've found. I'll push to this branch. Figured it'd be the most straightforward way for me just to show you what I've found.

**Comment by isaiahdahl** [comment]
@martijnversluis Pushed an update with a file full of tests. Going to start little discussions to highlight some things, or explain the changes. Then we can discuss it within the context of related code.

File: src/parser/chords_over_words_grammar.pegjs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+WordChar
+  = [^\n\r]
+
+ChordWithSpacing
+  = _ chord:Chord _ {
+      return chord;
+    }
+
+MetaData
```
Comment (isaiahdahl) [inline src/parser/chords_over_words_grammar.pegjs:81]: @martijnversluis  One thing you'll notice is changing it from frontMatter to metadata.

My reasoning for this was that we want to be able to support things like new_key and transpose. So basically metadata mid-song. And as soon as we start doing that, the word frontmatter didn't make sense to me anymore. So I thought it was simplest just to rename it to metadata.

File: src/parser/chords_over_words_grammar.pegjs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+          chordLyricsPairs.unshift({
+            type: "chordLyricsPair",
+            chord: null,
+            lyrics: lyrics.substring(0, firstChordPosition - 1),
+          });
+        }
+      }
+
+      return { type: "line", items: chordLyricsPairs };
+    }
```
Comment (bkeepers) [inline src/parser/chords_over_words_grammar.pegjs:51]: This looks good and is a similar approach to what I took in the [onsong](https://github.com/PraiseCharts/ChordSheetJS/pull/272/files#diff-53de6b9948a1bd529a53991dc298359b7b4ab4ca3c1f74361bb174a17947967cR22) parser. You could maybe eliminate the duplicate logic by unshifting an empty chord onto the `chords` array if the first chord does not start at column 0 prior to mapping chords.

**Review by isaiahdahl — CHANGES_REQUESTED** [review]
Just realized I created this big review the day you posted this and I never submitted the review. Whoops!!!

**Comment by martijnversluis** [comment]
@isaiahdahl Looks like all the tests pass now. I'll check and process your feedback and try and get this merged.

File: src/parser/chord_grammar.pegjs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+Chord
```
[the remark this answers is not in the document]
Comment (martijnversluis) [reply src/parser/chord_grammar.pegjs:1]: It is, it would be nicer if the grammar supported import directives.

We can, however, save the combined grammar as a git-ignored file to fix IDE inspections.

**Comment by isaiahdahl** [comment]
@martijnversluis  Thanks for your hard work on this! I resolved all the conversations because all the tests pass. 

Let's merge and then I'm going to prune through it further and create some more advanced integration tests to test the going back and forth between chordpro and chords over words. 

Then if I find holes I'll open up a new PR with those tests and we can work from there.

**Comment by martijnversluis** [comment]
@isaiahdahl Sounds great, happy to finally merge this chunk 😄 

I’ll at least start a new branch to work on numerics & numerals. On another branch we can work on chords variations (regardless of being chord symbol, numeric or numeral).
